docs/decisions/0012-system-taxonomy-creation.rst

```diff
@@ -30,26 +31,28 @@ We have two ways to handle Tags creation and validation for System-defined Taxon
 
 **Hardcoded by fixtures/migrations**
 
-#. If the tags don't change over the time, you can create all on a fixture (e.g Languages). 
+#. If the tags don't change over the time, you can create all on a fixture (e.g Languages).
+   The ``pk`` of these instances must be negative.
 #. If the tags change over the time, you can create all on a migration. If you edit, delete, or add new tags, you should also do it in a migration.
 
-**Free-form tags**
+**Dynamic tags**
 
-This taxonomy depends on a core data model, but simplifies the creation of Tags by allowing free-form tags,
-but we can validate the tags using the ``validate_object_tag`` method. For example we can put the ``AuthorSystemTaxonomy`` associated with
-the ``User`` model and use the ``ID`` field as tags. Also we can validate if an ``User`` still exists or has been deleted over time.
+Closed Taxonomies that depends on a core data model. Ex. AuthorTaxonomy with Users as Tags
+
+#. Tags are created on the fly when new ObjectTags are added.
+#. Tag.external_id we store an identifier from the instance (eg. User.pk).
+#. Tag.value we store a human readable representation of the instance (eg. User.username).
+#. Resync the tags to re-fetch the value. 
```

Comment:
OK, I definitely like the approach described here ^^

But what I'm still caught up on is `ModelObjectTag`. Do we need to subclass ObjectTag and push this logic into it? From reading the approach described here, and from what we discussed, I assumed we'd happily just use the normal `ObjectTag` or even `ContentObjectTag`. Because we're creating `Tag` instances as needed, and the regular `ObjectTag` already works correctly with `Tag` instances in Taxonomies.

Of course you need some logic to:
* **list available tags** even before they've been created as `Tag` instances: e.g. list all known users or languages → this belongs in the Taxonomy subclass but I don't think it's even necessary for this project, because we never need to display the "available" tags of system taxonomies to users. The tags will get auto-created by the system and the system already knows what user/language is in use. There's no UI and no dropdown menu to show.
*  **auto-create `Tag` instances** when tagging an object, if they don't yet exist. Sure, but this logic could live anywhere - in the `edx-platform` code that's applying the User and Language tags, in the `tag_object()` API, in the `Taxonomy.tag_object()` API - I don't think it is necessary to put this into an `ObjectTag` subclass, and doing so creates a lot of complexity.
   * Think MVP, and what's the simplest solution? Just auto-create the tag in the same place you apply the tag to some content. No special behavior needed in this repo at all. In fact I almost wonder if the User Taxonomy is just a regular closed taxonomy that's not editable, and the platform code can create `Tag`s in it just-in-time... What other functionality do we need here? 
* **auto-delete `Tag` instances** when the upstream object is deleted: we don't really need to worry about this for now, as languages won't be deleted and user deletion is very rare. In fact, I don't think user deletion even really happens; it's just retirement/anonymization that happens per GDPR. So in the future we could add a signal listener that detects "user retirement" and then deletes any `Tag` and `ObjectTag`s associated with the user, but I would leave that out of scope for now. We don't even really know how/where such user tags will be used yet.

Reply:
I agree with most of what you've said here, except for:

> Just auto-create the tag in the same place you apply the tag to some content. No special behavior needed in this repo at all.

I think providing a `ModelTaxonomy` (or mixin) that handles Tag auto-creation is broadly useful.

And I think it belongs in this openedx-learning library, because it's not content-specific and could be useful when tagging other types of objects, like people or forum posts. However, if it makes more sense for this to live in edx-platform for the MVP, I'm OK with that, we can always move it into openedx-learning later.

Reply:
> list available tags even before they've been created as Tag instances: e.g. list all known users or languages → this belongs in the Taxonomy subclass but I don't think it's even necessary for this project, because we never need to display the "available" tags of system taxonomies to users. The tags will get auto-created by the system and the system already knows what user/language is in use. There's no UI and no dropdown menu to show.

That's true except for language, it's the only [system-defined taxonomy that can be "edited"](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.uj6l8t9yvrwn) in a content by the content author.

> I think providing a ModelTaxonomy (or mixin) that handles Tag auto-creation is broadly useful.

I agree with Jill too, in addition to the auto creation, it checks that the instance (User, Organization) exists to see if the tag is valid.

> auto-delete Tag instances when the upstream object is deleted: we don't really need to worry about this for now, as languages won't be deleted and user deletion is very rare. In fact, I don't think user deletion even really happens; it's just retirement/anonymization that happens per GDPR. So in the future we could add a signal listener that detects "user retirement" and then deletes any Tag and ObjectTags associated with the user, but I would leave that out of scope for now. We don't even really know how/where such user tags will be used yet.

I agree, the simplest thing is to check if the user or organization exists. 

To clarify: the ModelTaxonomy is not used for languages, it is only used for the taxonomy of users and organizations

Reply:
> I think providing a ModelTaxonomy (or mixin) that handles Tag auto-creation is broadly useful.

👍🏻 Sure, good point. That's totally reasonable, and that sounds like a mixin to me. After all, in the platform if we keep the "ContentTaxonomy" vs "regular Taxonomy" distinction, it would have to be a mixin, to allow Content tagging with auto-creation of tags.

> That's true except for language, it's the only [system-defined taxonomy that can be "edited"](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.uj6l8t9yvrwn) in a content by the content author.

Oh OK. I've asked for clarification on that because I heard the opposite. But in any case, it seems we agree that "list available tags" should live in the Taxonomy. (I think it should be similar to what you implemented in `LanguageTaxonomy.get_tags() -> List[Tag]` but we'll probably eventually need an API that returns tag strings/IDs, not `Tag` objects because they may not exist yet. Imagine doing an auto-complete to tag users and there are 1,000,000 users in the system - we don't want to create `Tag` objects for each user unless they're actually used as a tag. Or maybe that would be a separate `search_available_tags` API. In any case, future PR.)

> I agree with Jill too, in addition to the auto creation, it checks that the instance (User, Organization) exists to see if the tag is valid.

That's fine. I do think however that belongs in the Taxonomy, not the ObjectTag itself. Because it's mostly a question of validating the User/Org then creating the corresponding `Tag` instance. The ObjectTag can validate itself with the relation to the `Tag` and doesn't need any special logic to handle ongoing validation related to the User/Org. If necessary, the Taxonomy can have some code to keep the `Tag` instances in sync with the User/Org models, but I think we can leave that out of scope for now to keep it simple.

Reply:
> Or maybe that would be a separate search_available_tags API. In any case, future PR.)

Yep, agreed -- and note we don't need to search model tags in MVP (User and Organization system taxonomies are our only model system taxonomies, and neither are editable by the content authors).

> I do think however that belongs in the Taxonomy, not the ObjectTag itself...but I think we can leave that out of scope for now to keep it simple.

Cool, thanks for your clarifications here @bradenmacdonald . I'll note this on my "cleanup" task.